modules/blocks/chain.js

```diff
@@ -552,100 +539,183 @@ Chain.prototype.broadcastReducedBlock = function(reducedBlock, broadcast) {
+			transaction,
+			undoUnconfirmErr => {
+				if (undoUnconfirmErr) {
+					// Fatal error, memory tables will be inconsistent
+					library.logger.error('Failed to undo transactions', undoUnconfirmErr);
+					return setImmediate(reject, undoUnconfirmErr);
+				}
+				return setImmediate(resolve);
+			},
+			tx
+		);
+	});
+};
 
-									return self.asyncProcessExit(0, cb);
-								}
+/**
+ * Performs backward tick
+ * @param {Object} oldLastBlock - secondLastBlock
+ * @param {Object} previousBlock - block to delete
+ * @param {Object} tx - database transaction
+ */
+__private.backwardTickStep = function(oldLastBlock, previousBlock, tx) {
+	return new Promise((resolve, reject) => {
+		// Perform backward tick on rounds
+		// WARNING: DB_WRITE
+		modules.rounds.backwardTick(
+			oldLastBlock,
+			previousBlock,
+			backwardTickErr => {
+				if (backwardTickErr) {
+					// Fatal error, memory tables will be inconsistent
+					library.logger.error(
+						'Failed to perform backwards tick',
+						backwardTickErr
+					);
+					return setImmediate(reject, backwardTickErr);
+				}
+				return setImmediate(resolve);
+			},
+			tx
+		);
+	});
+};
 
-								return setImmediate(cb, null, previousBlock);
-							});
-						});
-					}
-				);
-			}
+/**
+ * deletes block and relevant transactions
+ * @param {Object} oldLastBlock - secondLastBlock
+ * @param {Object} tx - database transaction
+ */
+__private.deleteBlockStep = function(oldLastBlock, tx) {
+	return new Promise((resolve, reject) => {
+		// Delete last block from blockchain
+		// WARNING: Db_WRITE
+		self.deleteBlock(
+			oldLastBlock.id,
+			deleteBlockErr => {
+				if (deleteBlockErr) {
+					// Fatal error, memory tables will be inconsistent
+					library.logger.error('Failed to delete block', deleteBlockErr);
+					return setImmediate(reject, deleteBlockErr);
+				}
+				return setImmediate(resolve);
+			},
+			tx
 		);
-	}, cb);
+	});
+};
+
+/**
+ * Deletes last block, undo transactions, recalculate round.
+ *
+ * @param  {function} cb - Callback function
+ * @returns {function} cb - Callback function from params (through setImmediate)
+ * @returns {Object} cb.err - Error
+ * @returns {Object} cb.obj - New last block
+ */
+__private.popLastBlock = function(oldLastBlock, cb) {
+	let secondLastBlock;
+
+	library.db
+		.tx('Chain:deleteBlock', tx =>
+			__private
+				.loadSecondLastBlockStep(oldLastBlock.previousBlock, tx)
+				.then(res => {
+					secondLastBlock = res;
+					return Promise.mapSeries(
+						oldLastBlock.transactions.reverse(),
+						transaction =>
+							__private
+								.undoStep(transaction, secondLastBlock, tx)
+								.then(() => __private.undoUnconfirmStep(transaction, tx))
+					);
+				})
+				.then(() =>
+					__private.backwardTickStep(oldLastBlock, secondLastBlock, tx)
+				)
+				.then(() => __private.deleteBlockStep(oldLastBlock, tx))
+		)
+		.then(() => setImmediate(cb, null, secondLastBlock))
+		.catch(err => {
+			process.exit(0);
```

Comment:
This is not a normal process exit, and should be `process.exit(1)`.

Reply:
I've replicated the logic as it existed earlier. But it's good that you raised this point. 

Because I wanted to discuss on the usages of process.exit over here. Earlier, we were calling `process.exit` because we thought we've mutated the db state which we'll be unable to recover from. But since now everything is a part of db transaction, we automatically revert all the changes made to the database. That being said, if deleting block is one of the basic actions of the application, and it should always work. So, if, because of some apparent reason, we are unable to delete a block (Which we previously verified and saved in the database), then there is definitely something wrong with application or database state. 
So, how do you think the application should behave over here? Should we exit or should we just return an error? @nazarhussain @4miners @MaciejBaj @LucasIsasmendi.

Reply:
Basing on the internal discussion between me and @SargeKhan, we've discussed to leave `process.exit` - in case of difference between unconfirmed and confirmed state in `mem_accounts`, the node restart can restore the consistency, which could possibly lead to successful block deletion in the next attempt.

Reply:
And I agree with @vitaly-t, basing on exit code documentation of Node.js https://nodejs.org/api/process.html#process_process_exit_code, we should exit with code = 1 when an internal error occurred. @SargeKhan please change.